### ISSUE TITLE
[java] CheckResultSet: False positive for rs.next() ? x : y

### ISSUE THREAD

**Issue by fabian-hahn** [pmd/pmd#7063]
**Affects PMD Version:**  7.27

**Rule:** CheckResultSet
<https://docs.pmd-code.org/pmd-doc-7.27.0/pmd_rules_java_bestpractices.html#checkresultset>

**Description:**

Calling `java.sql.ResultSet#next()` in a ternary condition flags an error which it probably shouldn't.

**Code Sample demonstrating the issue:**

```java
    int getCount(int id) throws SQLException {
        ResultSetHandler<Integer> countHandler =
            (rs) -> rs.next()? rs.getInt(1) : 0;

        String sql = """
            select count(*)
              from TABLE
             where ID = ?
        """;
        return runner.query(sql, countHandler, id);
    }
```

**Expected outcome:**

PMD reports a violation at line 3, but that's wrong. That's a false positive.

**Running PMD through:** *Maven*

**Comment by UncleOwen** [comment]
Yes, that rule needs an overhaul. See also https://github.com/pmd/pmd/issues/5031 and https://github.com/pmd/pmd/issues/5209

**Comment by UncleOwen** [comment]
Note that this rule is a more complicated case than [UnusedReturnValue](https://docs.pmd-code.org/latest/pmd_rules_java_errorprone.html#unusedreturnvalue). It's not really about rs.next() either. It's about what you're allowed to do afterwards.

There's one set of functions (rs.next(), rs.previous(), rs.first(), rs.last(), rs.absolute(), rs.relative()) that move the cursor. The return values of these functions tell you whether you're allowed to use rs.getRow(), rs.getInt(), etc. at the new position. If the return value is true you are allowed, if it's false (or before moving the cursor at all), you're not allowed. That's what this rule is (or at least should be) about.

(There's also rs.beforeFirst() and rs.beforeLast() that always move the cursor into an invalid position, functions that query the current position, etc. - but before we think about those we should get the basics I outlined above correct.)

**Comment by fabian-hahn** [comment]
Thanks, @UncleOwen for the fast reply.  

Hmm, I do not completely understand.  

```java
int c = 0;
if (rs.next()) c = rs.getInt();     // A

c = rs.next() ? rs.getInt() : 0;    // B
```
_A_ works just fine.  _B_ flags a violation. Shouldn't they behave the same?

**Comment by UncleOwen** [comment]
Yes, they should behave the same. What I'm saying is, that in my opinion that rule has deeper issues than that.
